### PR TITLE
[FLINK-5118] Fix inconsistent numBytesIn/Out metrics

### PR DESCRIPTION
**Pull request by zentol**
This PR fixes an issue regarding inconsistent numBytesIn/Out metrics.

Given 2 subsequent tasks A->B the numBytesOut count of A was lower than the numBytesIn(Local/Remote) count of B by a huge margin, although they should be (nearly) identical.

The problem is that A was counting how much bytes the serialized records were using, whereas B was counting how large the ```Buffer```s were that it received. A ```Buffer``` contains the following data:
```
|size_of_R1|R1|size_of_R2|R2|...
```
where as R1/R2 are serialized records and the sizes denote the serialized length of their respective record.

So, while A was adding ```sizeOf(R1)```, B was adding ```sizeOf(size_of_R1) + sizeOf(R1)```, which obviously leads to inconsistent counts.

A was simply not accounting for the added bytes that size_of_RX were using, which is what this PR is fixing by adding ```4``` after the serialization of each record.

cc @uce

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpanningRecordSerializer.java [hunk 1]
```diff
@@ -99,7 +99,7 @@ public SerializationResult addRecord(T record) throws IOException {
 		this.lengthBuffer.putInt(0, len);
 		
 		if (numBytesOut != null) {
-			numBytesOut.inc(len);
+			numBytesOut.inc(len + 4);
```
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpanningRecordSerializer.java]: I think this warrants both a comment and a test.

**Comment by uce** [comment]
Good catch. I was wondering whether it is better to count in a consistent way, too, e.g. count the size of each buffer or each record on both A and B (now we have per record on the output side and per buffer on the input side).

**Comment by zentol** [comment]
Instead of counting bytesOut in the serializer we could do it one level up in the RecordWriter.

**Comment by zentol** [comment]
Then we could count per buffer. I'd rather not count per record since the deserializer are kind of complicated :(

**Comment by uce** [comment]
Yeah, I think counting the buffers is the better way to go.

**Comment by zentol** [comment]
@uce Could you take another look? I moved the bytesOut counter into the RecordWriter.

**Comment by xhumanoid** [comment]
@zentol 
what do you think about remove 
if (numBytesOut != null) {

and replace
 numBytesOut = metrics.getNumBytesOutCounter();

with

+ if (metrics.getNumBytesOutCounter() != null) {
+    numBytesOut = metrics.getNumBytesOutCounter();
+ } else {
+    numBytesOut = new NullCounter();
+ }

where NullCounter have empty implementation for every method,

prof:
we do null check in one place, because sometime we may forget to do it

cons:
sometimes we broke devirtualization and inlining for counter.inc(..) method

**Comment by zentol** [comment]
@xhumanoid The metrics returned by the TaskIOMetricGroup can't actually be null, so I wouldn't put too much thought into dealing with that case.

**Comment by xhumanoid** [comment]
@zentol I asked because you always check on null when you try writing to Counter 
or is it prevent uninitialized state?

**Comment by zentol** [comment]
The code checks for null since there is no *technical* contract that the returned value is null. They aren't strictly necessary, and are only meant to guard against programming errors in the metrics system.

Using a ```NullCounter``` would indeed do the same, would however introduce effectively dead code.

**Comment by StephanEwen** [comment]
+1 to this approach.

Some suggestions for small performance improvements, not specific to this case, but also applicable to other cases:
  - It may make sense to type the counters at that level to the "SimpleCounter", if we anyways provide them. The TaskIOMetricGroup could type its counters to "SimpleCounter" as well. Since that class it not final, and future mocks/checks can be implemented on that class as well
  - We can make sure that the field `counter` is always initialized by initially assigning a standalone SimpleCounter. That way we could drop the null checks in the code.

**Comment by zentol** [comment]
@StephanEwen I'll add the default initialization of ```counter``` while merging.

Typing the counters to a ```SimpleCounter``` would affect a lot of other classes though (especially since we can extend this to the ```OperatorIOMetricGroup```), so I would like to do that as part of another issue.

**Pull request closed without merging**